File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());

        for(auto i = 0; i < nums.size() - 2; i++) {
            auto left = i + 1;
            auto right = nums.size() - 1;

            while(left < right) {
                auto sum = nums[i] + nums[left] + nums[right];

                if (sum < 0) left++;
                else if (sum > 0) right--;
                else {
                    result.push_back(vector<int>{nums[i], nums[left], nums[right]});
                    while(left + 1 < right && nums[left] == nums[left + 1]) left++;
                    while(left < right - 1 && nums[right] == nums[right - 1]) right--;
                    left++, right--;
                }   
            }
            while(i + 1 < nums.size() - 2 && nums[i] == nums[i + 1]) i++;
        }
        return result;
    }
};

File: 15-3sum/3sum.go
func threeSum(nums []int) [][]int {
    n := len(nums)
    seen := make(map[[3]int]bool)
    result := make([][]int, 0)
    sort.Slice(nums, func(i, j int) bool {
        return nums[i] < nums[j]
    })

    for i, v := range nums {
        target := -v
        left, right := i + 1, n - 1

        for left < right {
            sum := nums[left] + nums[right]
            trip := [3]int{v, nums[left], nums[right]}
            if sum == target && !seen[trip] {
                result = append(result, []int{v, nums[left], nums[right]})
                seen[trip] = true
            }
            if sum > target {
                right--
            } else {
                left++
            }
        }
    }

    return result
}

